### PR TITLE
RFC28: Use Block Timestamp as Start Timestamp in Since

### PR DESCRIPTION
**Pull request by doitian**

### CODE REVIEW

**Comment by doitian** [comment]
@keroro520 fixed the inequality a1f3442d279f0e8178dda97c22eb92d6a2c63476

**Review by keroro520 — COMMENTED** [review]
LGTM

File: rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+where
+
+* `StartTime` is the block timestamp that committed the cell consumed by the input.
+* `SinceValue` is the `value` part of the `since` field.
+* `MedianTimestamp` is the median timestamp of the previous 37 blocks preceding the block if the transaction is in the block, or the latest 37 blocks if the transaction is in the pool.
+
+The transaction verification fails if the transaction is immature.
+
+The only change is `StartTime`, which was the median of the previous 37 blocks preceding the one that has committed the consumed cell. Because block timestamp must be larger than the median of its previous 37 blocks, the new consensus rule is stricker than the old rule. The transaction that is mature using the old rule may be immature using the new rule, but the transaction that is mature using the new rule must be mature using the old rule.
```
Comment (yangby-cryptape) [inline rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md:46]:
```suggestion
The only change is `StartTime`, which was the median of the previous 37 blocks preceding the one that has committed the consumed cell. Because block timestamp must be larger than the median of its previous 37 blocks, the new consensus rule is stricter than the old rule. The transaction that is mature using the old rule may be immature using the new rule, but the transaction that is mature using the new rule must be mature using the old rule.
```

File: rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+The deployment can be performed in two stages.
+
+The first stage will activate the new consensus rule starting from a specific epoch. The mainnet and testnet will use different starting epochs and all other chains will use the new rule from epoch 0.
+
+After the fork is activated, and if the transactions in the old rule all satisfies the new rule, the old consensus rule will be removed and the new rule will be applied from the genesis block.
+
+## Backward compatibility
+
+Because the new consensus rule is stricker than the old one, this proposal can be deployed via a soft fork.
```
Comment (yangby-cryptape) [inline rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md:76]:
```suggestion
Because the new consensus rule is stricter than the old one, this proposal can be deployed via a soft fork.
```

File: rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	* The `metric_flag` is block timestamp (10).
+	* The `relative_flag` is relative (1).
+	* The `value` is 600,000 (10 minutes).
+* The median of the previous 37 blocks preceding block S is 10,000.
+* The timestamp of block S is 20,000.
+* The median of the previous 37 blocks preceding block T is 615,000
+
+In the old consensus, StartTime + SinceValue = 10,000 + 600,000 = 610,000, which is less than the MedianTimestamp 615,000, thus the transaction is mature.
+
+But in the new rule, StartTime + SinceValue = 20,000 + 600,000 = 620,000 ≥ 615,000, so the transactino is still immature.
```
Comment (yangby-cryptape) [inline rfcs/0221-change-since-relative-timestamp/0221-change-since-relative-timestamp.md:64]:
```suggestion
But in the new rule, StartTime + SinceValue = 20,000 + 600,000 = 620,000 ≥ 615,000, so the transaction is still immature.
```

**Review by yangby-cryptape — CHANGES_REQUESTED** [review]

**Review by keroro520 — APPROVED** [review]

**Review by yangby-cryptape — DISMISSED** [review]

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 4]
```diff
@@ -1 +1,76 @@
...
+
+## Abstract
+
+This document proposes a transaction verification consensus change. When the `since` field of the transaction input uses a relative timestamp, the referenced cell committing block timestamp is used instead of the median timestamp.
+
+This is a modification to the RFC17 [Transaction valid since](../0017-tx-valid-since/0017-tx-valid-since.md).
+
+## Motivation
+
+The current consensus rule uses the median of the timestamp fields of the 37 blocks preceding the referenced cell committing block. It is resource consuming to get the median timestamp because it requires either getting 37 block headers or caching the median timestamp for each block. The intention to use the median time is to avoid miners profitable by manipulating block timestamp to include more transactions, but it is safe to use the committing block timestamp as the start time because of two reasons:
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:20]:
```suggestion
The current consensus rule uses the median of the timestamps in the 37 blocks preceding the referenced cell committing block. Getting the median timestamp is resource consuming because it requires either getting 37 block headers or caching the median timestamp for each block. The intention of using median time was to prevent miners from manipulating block timestamp to include more transactions. But it is safe to use the committing block timestamp as the start time because of two reasons:
```

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 5]
```diff
@@ -1 +1,76 @@
...
+
+1. The timestamp in the block header has already been verified by the network that it must be larger than the median of the previous 37 blocks and less than or equal to the current time plus 15 seconds. TODO: reference to block data structure RFC
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:22]: Could you add the reference link and remove TODO?

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 6]
```diff
@@ -1 +1,76 @@
...
+
+where
+
+* `StartTime` is the block timestamp that commits the cell consumed by the input.
+* `SinceValue` is the `value` part of the `since` field.
+* `MedianTimestamp` is the median timestamp of the previous 37 blocks preceding the block if the transaction is in the block, or the latest 37 blocks if the transaction is in the pool.
+
+The transaction verification fails if the transaction is immature.
+
+The only change is `StartTime`, which was the median of the previous 37 blocks preceding the one that has committed the consumed cell. Because block timestamp must be larger than the median of its previous 37 blocks, the new consensus rule is more strict than the old rule. The transaction that is mature using the old rule may be immature using the new rule, but the transaction that is mature using the new rule must be mature using the old rule.
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:46]:
```suggestion
The only change is `StartTime`, which was the median of the previous 37 blocks preceding the one that has committed the consumed cell. Because block timestamp must be larger than the median of its previous 37 blocks, the new consensus rule is more strict than the old rule. A transaction that is mature under the old rule may be immature under the new rule, but a transaction that is mature under the new rule must be mature under the old rule.
```

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 7]
```diff
@@ -1 +1,76 @@
...
+
+* A transaction consumes a cell in block S and is about to be committed into block T with a since requirement that:
+	* The `metric_flag` is block timestamp (10).
+	* The `relative_flag` is relative (1).
+	* The `value` is 600,000 (10 minutes).
+* The median of the previous 37 blocks preceding block S is 10,000.
+* The timestamp of block S is 20,000.
+* The median of the previous 37 blocks preceding block T is 615,000
+
+In the old consensus, StartTime + SinceValue = 10,000 + 600,000 = 610,000, which is less than the MedianTimestamp 615,000, thus the transaction is mature.
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:62]:
```suggestion
In the old consensus, `StartTime` + `SinceValue` = 10,000 + 600,000 = 610,000, which is less than the `MedianTimestamp` 615,000, thus the transaction is mature.
```

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 8]
```diff
@@ -1 +1,76 @@
...
+
+But in the new rule, StartTime + SinceValue = 20,000 + 600,000 = 620,000 ≥ 615,000, so the transaction is still immature.
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:64]:
```suggestion
But in the new rule, `StartTime` + `SinceValue` = 20,000 + 600,000 = 620,000 ≥ 615,000, so the transaction is still immature.
```

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 9]
```diff
@@ -1 +1,76 @@
...
+
+This is a modification to the RFC17 [Transaction valid since](../0017-tx-valid-since/0017-tx-valid-since.md).
+
+## Motivation
+
+The current consensus rule uses the median of the timestamp fields of the 37 blocks preceding the referenced cell committing block. It is resource consuming to get the median timestamp because it requires either getting 37 block headers or caching the median timestamp for each block. The intention to use the median time is to avoid miners profitable by manipulating block timestamp to include more transactions, but it is safe to use the committing block timestamp as the start time because of two reasons:
+
+1. The timestamp in the block header has already been verified by the network that it must be larger than the median of the previous 37 blocks and less than or equal to the current time plus 15 seconds. TODO: reference to block data structure RFC
+2. The transaction consuming a cell with the `since` requirement is committed later than the transaction creating the cell. When miners see the consuming transaction, the start timestamp has already been determined.
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:23]: Could you rephrase and make the point more clear?

**Review by janx — CHANGES_REQUESTED** [review]

**Review by yangby-cryptape — DISMISSED** [review]

File: rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md [hunk 10]
```diff
@@ -1 +1,76 @@
-In review, see <https://github.com/nervosnetwork/rfcs/pull/221>
+---
+Number: "0028"
+Category: Consensus (Soft Fork)
```
Comment (janx) [inline rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md:3]: Category must be in either `Standards Track` or `Informational`. Otherwise a new category with its description should be added first.

I prefer to move forward with existing category first. After new RFCs merged (will be in draft status) we can come back to the category issue and see if we need to update RFC categories.

**Review by janx — CHANGES_REQUESTED** [review]

**Review by yangby-cryptape — APPROVED** [review]

**Review by janx — APPROVED** [review]

**Review by quake — APPROVED** [review]
